server/handlers.go

```diff
@@ -749,12 +759,17 @@ func (s *Server) handleToken(w http.ResponseWriter, r *http.Request) {
 		}
 		return
 	}
-	if client.Secret != clientSecret {
+
+	grantType := r.PostFormValue("grant_type")
+	codeVerifier := r.PostFormValue("code_verifier")
+
+	if grantType == grantTypeAuthorizationCode && codeVerifier != "" {
```

Comment:
Is this really necessary? What's wrong with checking clientSecret even for PKCE?

Reply:
If we would check the clientSecret for PKCE, the client would have to possess the client secret. PKCE is created for public clients that **cannot safely** store a clientSecret; e.g. single-page applications.

[Very good explanation of Auth Code flow with PKCE](https://auth0.com/docs/flows/authorization-code-flow-with-proof-key-for-code-exchange-pkce)

However; if someone does not want to have PKCE enabled, because he **always** wants to use the client secret, we should think about a configuration to enable/disable PKCE.

Reply:
I agree that PKCE was introduced to mitigate security concerns for public clients, but my reasoning is: 
If client is configured as public, then the check should pass as both clientSecret and client.Secret should be empty.
But if for whatever reason sb configures confidential client with PKCE, why not validate client secret as well.

Reply:
Dex currently disallows an empty client secret with the error:
`invalid config: Secret or SecretEnv field is required for client "example-app"`
and I don't want to mess with that decision. The secret protects e.g. the refresh_token flow and the password_grant flow.

I guess it is best to have a config parameter for the client like
`allowPKCE: true` which is false by default. This way, old configurations still use client_secret only and are not affected by the PKCE implementation.

Sounds good?

Reply:
With "public" you mean setting the client.public parameter and thus enabling https://github.com/dexidp/dex/pull/471?

### Suggestion one
```suggestion
        if client.Public && grantType == grantTypeAuthorizationCode && codeVerifier != "" {
               // let it pass
        } else // check secret
```
This will accomplish the following:

if client is confidential, the clientSecret will always be checked (
if client is public, but the client has a client secret, PKCE will still work *without* a client secret.

### Suggestion two
introduce a new boolean to client configuration, which allows PKCE without checking the client secret. This way the client does not have to be set to public.

Reply:
`With "public" you mean setting the client.public parameter and thus enabling #471?`
Yes.

I am trying to understand the scenario you have in mind behind the this if-check. Are you concerned about allowing access in scenario of having a single client configuration dex-side (as confidential client) and then have 2 clients that use it - one is confidential (knows the secret and can protect it) and other is public (does not know the secret and uses PKCE)?

Reply:
> Are you concerned about allowing access in scenario of having a single client configuration dex-side (as confidential client) and then have 2 clients that use it - one is confidential (knows the secret and can protect it) and other is public (does not know the secret and uses PKCE)?

Yes.

Reply:
I think this use-case might violate the OIDC spec:
https://openid.net/specs/openid-connect-core-1_0.html
```
9.  Client Authentication
This section defines a set of Client Authentication methods that are used by Clients to authenticate to the Authorization Server when using the Token Endpoint. During Client Registration, the RP (Client) MAY register a Client Authentication method. If no method is registered, the default method is client_secret_basic.

These Client Authentication methods are:

client_secret_basic
Clients that have received a client_secret value from the Authorization Server authenticate with the Authorization Server in accordance with Section 2.3.1 of OAuth 2.0 [RFC6749] using the HTTP Basic authentication scheme.
[...]
none
The Client does not authenticate itself at the Token Endpoint, either because it uses only the Implicit Flow (and so does not use the Token Endpoint) or because it is a Public Client with no Client Secret or other authentication mechanism.
```

This seems to imply that there should be a single authentication method used by the client and your scenario assumes two (some client instances would use `client_secret_basic`, others `none`)

Reply:
Also reading through [this](https://github.com/mitreid-connect/OpenID-Connect-Java-Spring-Server/issues/1456) and [this](https://stackoverflow.com/questions/56793968/should-authorization-flow-use-both-secret-and-pkce-for-a-web-app), I agree upon that. Will change it.

Reply:
Four pur application, the restriction of the callback uris implied by the setting "public" does not work out.

Reply:
Actually, i don't quite understand, why "public" clients don't respect the configured redirectURIs.

Reply:
The document even states that wildcard domains should be prevented. The current "private" implementation violates this by basically implementing redirect URI of type `http[s]://localhost/*`.
But implementing it according to that spec, would mean to possibly break existing dex deployments :/

Reply:
I think i misread it, so i mimiized above comment. Should I implement the usage of RedirectURI in "private" in this pull request, because PKCE does not work correctly for SPAs without it, or should that be discussed in #1300?

Reply:
when you say "private" you mean "public"? I think this should be discussed in #1300. I wonder if there are any downsides to just using confidential client with dummy password as a workaround for your case, don't see any at first glance.
Also, I guess this would be a good time to get some actual maintainer to interested in this PR

Reply:
> when you say "private" you mean "public"?

yes. (D'oh!)
> I think this should be discussed in #1300. I wonder if there are any downsides to just using confidential client with dummy password as a workaround for your case, don't see any at first glance.

The SPA draft states `NOT RECOMMENDED` to use a client_secret in that case. I am worried that we would carry that useless client_secret around "forever".
> Also, I guess this would be a good time to get some actual maintainer to interested in this PR

definitively!